mlflow/tracing/attachments.py

```diff
@@ -0,0 +1,183 @@
+            f"size={len(self.content_bytes)} bytes)"
+        )
+
+    @classmethod
+    def from_file(cls, file: str | Path, content_type: str | None = None) -> "Attachment":
+        """Create an Attachment from a file path.
+
+        Args:
+            file: Path to the file to attach
+            content_type: Optional content type override. If not provided, will be inferred from
+                file extension.
+
+        Returns:
+            Attachment object
+        """
+        path = Path(file)
+        return cls(
+            content_type=content_type or cls._infer_content_type(file),
+            content_bytes=path.read_bytes(),
+            filename=path.name,
+        )
+
+    @staticmethod
+    def _infer_content_type(file: str | Path) -> str:
+        """Infer content type from file extension."""
+        # Use mimetypes module for more comprehensive type detection
+        content_type, _ = mimetypes.guess_type(file)
+        if content_type:
+            return content_type
+
+        # Fallback to manual mapping for common types
+        file_lower = file.lower()
+        if file_lower.endswith((".png",)):
+            return "image/png"
+        elif file_lower.endswith((".jpg", ".jpeg")):
+            return "image/jpeg"
+        elif file_lower.endswith((".gif",)):
+            return "image/gif"
+        elif file_lower.endswith((".webp",)):
+            return "image/webp"
+        elif file_lower.endswith((".pdf",)):
+            return "application/pdf"
+        elif file_lower.endswith((".mp3",)):
+            return "audio/mpeg"
+        elif file_lower.endswith((".wav",)):
+            return "audio/wav"
+        elif file_lower.endswith((".ogg",)):
+            return "audio/ogg"
+        elif file_lower.endswith((".m4a",)):
+            return "audio/mp4"
+        elif file_lower.endswith((".aac",)):
+            return "audio/aac"
+        elif file_lower.endswith((".flac",)):
+            return "audio/flac"
+        elif file_lower.endswith((".webm",)):
+            return "audio/webm"
+        elif file_lower.endswith((".mp4",)):
+            return "video/mp4"
+        elif file_lower.endswith((".mov",)):
+            return "video/quicktime"
+        elif file_lower.endswith((".avi",)):
+            return "video/x-msvideo"
+        return "application/octet-stream"
+
+    def ref(self, trace_id: str, artifact_uri: str | None = None) -> str:
+        """
+        A string representation of the attachment reference.
+
+        Args:
+            trace_id: The trace ID this attachment belongs to
+            artifact_uri: The artifact URI where the trace is stored.
+                If None, uses a default placeholder.
+
+        Returns:
+            JSON-based reference string encoded in base64 for URL safety
+        """
+        metadata = {
+            "attachment_id": self.id,
+            "trace_id": trace_id,
+            "artifact_uri": artifact_uri or "mlflow-artifacts:/",
```

Comment:
`artifact_uri` is included so this object can self-fetch the file without fetching the trace.